curl_fuzzer_tlv_bufq.cc

```diff
@@ -0,0 +1,141 @@
+/***************************************************************************
+ *                                  _   _ ____  _
+ *  Project                     ___| | | |  _ \| |
+ *                             / __| | | | |_) | |
+ *                            | (__| |_| |  _ <| |___
+ *                             \___|\___/|_| \_\_____|
+ *
+ * Copyright (C) 2017, Max Dymond, <cmeister2@gmail.com>, et al.
+ *
+ * This software is licensed as described in the file COPYING, which
+ * you should have received as part of this distribution. The terms
+ * are also available at https://curl.se/docs/copyright.html.
+ *
+ * You may opt to use, copy, modify, merge, publish, distribute and/or sell
+ * copies of the Software, and permit persons to whom the Software is
+ * furnished to do so, under the terms of the COPYING file.
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTY OF ANY
+ * KIND, either express or implied.
+ *
+ ***************************************************************************/
+#include <stdlib.h>
+#include <string.h>
+#include <curl/curl.h>
+#include "curl_fuzzer.h"
+
+/**
+ * Do different actions on the CURL handle for different received TLVs.
+ */
+int fuzz_parse_tlv(FUZZ_DATA *fuzz, TLV *tlv)
+{
+  int rc;
+  OPERATION *tmp_op;
+  uint32_t tmp_op_type;
+  uint32_t tmp_u32;
+
+  switch(tlv->type) {
+    case TLV_TYPE_MAX_CHUNKS:
+      FCHECK(tlv->length == 4);
+      FCHECK(fuzz->max_chunks == 0);
+      tmp_u32 = to_u32(tlv->value);
+      FCHECK(tmp_u32 > 0 && tmp_u32 < TLV_MAX_CHUNKS_QTY);
+      fuzz->max_chunks = tmp_u32;
+      break;
+    
+    case TLV_TYPE_CHUNK_SIZE:
+      FCHECK(tlv->length == 4);
+      FCHECK(fuzz->chunk_size == 0);
+      tmp_u32 = to_u32(tlv->value);
+      FCHECK(tmp_u32 > 0 && tmp_u32 < TLV_MAX_CHUNK_SIZE);
+      fuzz->chunk_size = tmp_u32;
+      break;
+
+    case TLV_TYPE_MAX_SPARE:
+      FCHECK(tlv->length == 4);
+      FCHECK(fuzz->max_spare == 0);
+      tmp_u32 = to_u32(tlv->value);
+      FCHECK(tmp_u32 > 0 && tmp_u32 < TLV_MAX_MAX_SPARE);
+      fuzz->max_spare = tmp_u32;
+      break;
+
+    case TLV_TYPE_USE_POOL:
+      FCHECK(tlv->length == 0);
+      FCHECK(fuzz->use_pool == 0);
+      fuzz->use_pool = 1;
+      break;
+
+    case TLV_TYPE_NO_SPARE:
+      FCHECK(tlv->length == 0);
+      FCHECK(fuzz->no_spare == 0);
+      fuzz->no_spare = 1;
+      break;
+
+    case TLV_TYPE_RESET:
+      tmp_op_type = OP_TYPE_RESET;
+      goto ADD_OP_UNSIZED;
```

Comment:
Not convinced about using `goto`s here.

I think switching out this whole mechanism for the FuzzedDataProvider makes more sense given the classes of operations you're trying to do here. With that we could in a loop do:

- get a byte for the operation type
- get extra bytes if required for the operation
- operate directly on the buffer (no creating operation lists)
- continue until FuzzedDataProvider is empty.

It'd also be a smidge faster because you're not having to precalculate the operations list (and allocating memory for it).

Reply:
Hi @cmeister2, I reverted the TLV changes and reworked this with FuzzedDataProvider, let me know if that looks any better 👍  execs/s as reported by libfuzzer are significantly reduced, but I assume the TLV approach had many more 'rejected' executions that are shorter.

Reply:
This looks awesome, thanks for implementing it! I don't know the internal mechanics of FuzzedDataProvider but I hope it's optimised for this kind of usage.